### ISSUE TITLE
False singleton dimension for scalar targets considered harmful

### ISSUE THREAD

**Issue by dwf** [mila-udem/fuel#137]
Consider this code:

**Comment by vdumoulin** [comment]
@dwf is this issue still relevant?

**Comment by lamblin** [comment]
I think it is, since it also makes it trickier to adapt non-Fuel code to use Fuel, as I experienced when adapting the tutorials for the deep learning summer school.

**Comment by dwf** [comment]
Indeed, it's kind of annoying to remember to flatten each time. I also
don't totally buy the metadata argument that was the justification: that
the dimension label should be "index". That is not, in fact, the label for
the dimension but a label for a particular column on that dimension. What
you're describing with that is more the semantics of elements of the
"space" in which the items lie.

**Comment by bartvm** [comment]
I personally like the consistency that it is a 2D array, because it makes having a single label just a special case of the multiple label case i.e. `batch[i]` returns an array of labels for a sample. In this case that array has length 1 (i.e. `batch` is a column vector), but that doesn't always have to be the case, and `batch` could just as well be a 2D array or a nested list.

**Comment by dwf** [comment]
If we're already giving up on the elements `batch[i]` being arrays (i.e. they could be lists), then I don't care if they're sequences or not. The tradeoff of using `atleast_1d()` in places where I expect an array is better than accidental silent broadcasting, as operations that explicitly require a sequence (or more specifically an array) and receive a scalar will *usually* raise an error, whereas operations that expect a vector and get a matrix will often not (as demonstrated by the example in the description text).
